### ISSUE TITLE
Bitch of a time running Stan in Julia

### ISSUE THREAD

**Issue by gregbritten** [jpmattern/bayesian_cbiomes.pg#1]
Wu (lab mate) and I tried installing and running Stan in Julia yesterday and had a bitch of a time. It's still not working on my PC. Wu was able to get it to work on his Mac after having to fix several weird paths and settings. (I remember folks in the past having a harder time with Julia). Could you give it a go on your system and let me know if you run into issues? We'll have to sort out potential issues if we're going to support Julia in the workshop.. 
I'm going to give it another go on my PC and will post a question to the author of CmdStan.jl/Stan.jl
I'm also going to try it on my old Mac to see how it goes.

**Comment by jpmattern** [comment]
Alright, trying it out on two Linux machines (office computer running Julia 1.0.3 and cluster main node running Julia 1.1.1).

On office computer I am having a bad start:
```
julia> using Stan
[ Info: Precompiling Stan [682df890-35be-576f-97d0-3d8c8b33a550]
ERROR: LoadError: No deps.jl file could be found. Please try running Pkg.build("Arpack").
Currently, the build command might fail when Julia has been built from source
and the recommendation is to use the official binaries from julialang.org.
For more info see https://github.com/JuliaLinearAlgebra/Arpack.jl/issues/5.
```
On the cluster I am progressing a bit farther bu then I run into this error:
```
ERROR: The environment variable JULIA_CMDSTAN_HOME needs to be set.
```
... which tell me that cmdstan is needed to run Stan and I cannot do that on the cluster computer (though I could perhaps request it).
BTW, I was trying to run [this example](https://github.com/StanJulia/Stan.jl/blob/master/Examples/Bernoulli/bernoulli.jl)

**Comment by gregbritten** [comment]
Indeed. CmdStan.jl and Stan.jl (both seem to do the same thing - I think perhaps Stan.jl is the new package still in development) are just Julia interfaces for cmdstan which is a package to run Stan at command line. 

For whatever reason I cannot get CmdStan.jl to start cmdstan.. I fix one error and another pops up.. over and over...

**Comment by jpmattern** [comment]
I tried again and success! I downloaded the latest binary of Julia (version 1.2.0) and used that. But I needed to start in bash:
 ```
cd /some/path
git clone https://github.com/stan-dev/cmdstan.git --recursive
cd cmdstan
make build
export JULIA_CMDSTAN_HOME=/some/path/cmdstan
```
Now in Julia:
```
(v1.2) pkg> add Stan
(v1.2) pkg> add Arpack
(v1.2) pkg> build Arpack
julia> using Arpack
[ Info: Precompiling Arpack [7d9fca2a-8960-54d3-9f78-7d1dccf2cb97]
julia> using Stan
[ Info: Precompiling Stan [682df890-35be-576f-97d0-3d8c8b33a550]
[ Info: Loading DataFrames support into Gadfly.jl
(v1.2) pkg> add StatsPlots
(v1.2) pkg> add StanSample
```
Finally, and I am not sure if this step is needed but I did it:
```
julia> using StanSample
julia> StanSample.set_cmdstan_home!("/some/path/cmdstan/")
```
After that I was able to run the example.

**Comment by jpmattern** [comment]
Hmmm... one odd complication is apparently that "Stan" requires `CMDSTAN_HOME` to be set, while  "StanSample" relies on `JULIA_CMDSTAN_HOME`. Am I correct in assuming, that we'll be using StanSample and that Stan is some future version?

**Comment by gregbritten** [comment]
Great to know that you had relatively minimal issues.. although Arpack was a new twist. 
My PC is still stuck, but no problems on my Mac (after setting CMDSTAN_HOME and JULIA_CMDSTAN_HOME). I still don't know why I opted for a PC laptop... :-S
Anyway, I will keep working to see if I can isolate the PC issue, and I will try on another PC to see if the headaches appear consistently.. and whether we will be able to give the group a robust set of instructions for the PC/Julia combination. 

As for StanSample.jl vs. Stan.jl, I'm not too sure, because I see StanSample examples on the Stan.jl GitHub.. and I also see the authors note that big changes are coming to Stan.jl so perhaps you are right that it is a work in progress, and perhaps StanSample.jl will become part of Stan.jl. CmdStan.jl seems to be the older and more common when I Google around, so we may want to try working with that too..

I'll keep the issue open until I have more PC updates. For now I am going to continue working on the introduction to Bayes/Stan notebook as that is more fun and less frustrating..

**Comment by jpmattern** [comment]
Well, good to hear you got it working on the Mac at least, i think the combo Windows+Julia might be pretty rare, though we should have probably asked that in the survey. We could still send out an email asking everyone about their OS...

**Comment by gaelforget** [comment]
Not sure if this will help (and maybe it's already been discussed above) but one possibility could be to use an external installation (done via Conda or pip) through `PyCall` / `pyImport`.

That approach has worked for me in order to call `pycmap` in `julia` (see `DarwinCmapJulia.ipynb` in https://github.com/gaelforget/Cbiomes2019Notebooks) and also recently to use `Cartopy` in `Julia` (see `src/PlotIndDisp.jl` in https://github.com/gaelforget/IndividualDisplacements.jl). 

Just in case this is useful...

**Comment by gaelforget** [comment]
Please edit the issue title to something like `Tough time running Stan in Julia`. Looking at the current title of this issue in my GitHub activity feed makes me incomfortable.

_Please, as soon as you get a chance_ 

Thanks @jpmattern for your prompt intervention

**Comment by jpmattern** [comment]
Haha, edited the title. I, too, got confused when I had a tab open that said "Bitch".

Gael, would you say your approach is easy enough for a bunch of participants to copy? You are using a mac, I assume?

**Comment by gregbritten** [comment]
Interesting, @gaelforget so you are suggesting that we interact with PyStan from Jupyter via PyCall? 
That may end up saving us some headaches because PyStan is quite a bit better developed than Stan.jl/CmdStan.jl/StanSample.jl which are all various wrappers for running Stan at command line outside of Julia. This intermediate wrapper step seems to be a bit finicky... 
If PyCall works well, then I think we would benefit from the greater robustness of PyStan.

**Comment by gregbritten** [comment]
And my apologies for making you guys uncomfortable with my bad language :-P

**Comment by gregbritten** [comment]
I am also noticing an RCall package!

**Comment by gaelforget** [comment]
> would you say your approach is easy enough for a bunch of participants to copy? 

Don't know -- calling c++ from Julia via python or R is probably never going to be _easy enough_

> You are using a mac, I assume?

yes although my `DarwinCmapJulia.ipynb` example also has worked using a binder cloud instance which ran in linux I think (you can try via the badge; please post an issue to the repo if that fails)

> are suggesting that we interact with PyStan from Jupyter via PyCall?

Not sure but figured you might want to give it a try
